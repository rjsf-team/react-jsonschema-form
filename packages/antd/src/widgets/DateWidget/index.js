import React from 'react';
import dayjs from 'dayjs';

import DatePicker from '../../components/DatePicker';

const DATE_PICKER_STYLE = {
  width: '100%',
};

const DateWidget = ({
  // autofocus,
  disabled,
  formContext,
  id,
  // label,
  onBlur,
  onChange,
  onFocus,
  // options,
  placeholder,
  readonly,
  // required,
  // schema,
  value,
}) => {
  const { readonlyAsDisabled = true } = formContext;

  const handleChange = (nextValue) =>
    onChange(nextValue && nextValue.format('YYYY-MM-DD'));

  const handleBlur = () => onBlur(id, value);

  const handleFocus = () => onFocus(id, value);

  const getPopupContainer = (node) => node.parentNode;

  return (
    <DatePicker
      disabled={disabled || (readonlyAsDisabled && readonly)}
      getPopupContainer={getPopupContainer}
      id={id}
      name={id}
      onBlur={!readonly ? handleBlur : undefined}
      onChange={!readonly ? handleChange : undefined}
      onFocus={!readonly ? handleFocus : undefined}
      placeholder={placeholder}
      showTime={false}
      style={DATE_PICKER_STYLE}
      value={value && dayjs(value)}
    />
  );
};

export default DateWidget;
