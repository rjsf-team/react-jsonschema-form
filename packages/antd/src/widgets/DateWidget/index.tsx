import dayjs from 'dayjs';
import {
  ariaDescribedByIds,
  FormContextType,
  RJSFSchema,
  StrictRJSFSchema,
  WidgetProps,
  GenericObjectType,
} from '@rjsf/utils';

import DatePicker from '../../components/DatePicker';

const DATE_PICKER_STYLE = {
  width: '100%',
};

/** The `DateWidget` component uses the `BaseInputTemplate` changing the type to `date` and transforms
 * the value to undefined when it is falsy during the `onChange` handling.
 *
 * @param props - The `WidgetProps` for this component
 */
export default function DateWidget<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WidgetProps<T, S, F>
) {
  const { disabled, formContext, id, onBlur, onChange, onFocus, placeholder, readonly, value } = props;
  const { readonlyAsDisabled = true } = formContext as GenericObjectType;

  const handleChange = (nextValue: any) => onChange(nextValue && nextValue.format('YYYY-MM-DD'));

  const handleBlur = () => onBlur(id, value);

  const handleFocus = () => onFocus(id, value);

  const getPopupContainer = (node: any) => node.parentNode;

  return (
    <DatePicker
      disabled={disabled || (readonlyAsDisabled && readonly)}
      getPopupContainer={getPopupContainer}
      id={id}
      name={id}
      onBlur={!readonly ? handleBlur : undefined}
      onChange={!readonly ? handleChange : undefined}
      onFocus={!readonly ? handleFocus : undefined}
      placeholder={placeholder}
      showTime={false}
      style={DATE_PICKER_STYLE}
      value={value && dayjs(value)}
      aria-describedby={ariaDescribedByIds<T>(id)}
    />
  );
}
