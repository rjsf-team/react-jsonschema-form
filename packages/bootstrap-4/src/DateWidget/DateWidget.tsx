import React from "react";
import { WidgetProps } from '@rjsf/core';

const DateWidget = (props: WidgetProps) => {
  const { registry } = props;
  const { TextWidget } = registry.widgets;
  return (
    <TextWidget
      {...props}
      type="date"
    />
  );
};

export default DateWidget;
