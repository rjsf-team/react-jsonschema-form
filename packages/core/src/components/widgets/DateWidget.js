import React from "react";
import PropTypes from "prop-types";

function DateWidget(props) {
  const {
    onChange,
    registry: {
      widgets: { BaseInput },
    },
  } = props;
  return (
    <BaseInput
      type="date"
      {...props}
      onChange={(value) => onChange(value || undefined)}
    />
  );
}

if (process.env.NODE_ENV !== "production") {
  DateWidget.propTypes = {
    value: PropTypes.string,
  };
}

export default DateWidget;
