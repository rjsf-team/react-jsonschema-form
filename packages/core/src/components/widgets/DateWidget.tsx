import { useCallback } from 'react';
import { getTemplate, FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';

/** The `DateWidget` component uses the `BaseInputTemplate` changing the type to `date` and transforms
 * the value to undefined when it is falsy during the `onChange` handling.
 *
 * @param props - The `WidgetProps` for this component
 */
export default function DateWidget<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WidgetProps<T, S, F>
) {
  const { onChange, options, registry } = props;
  const BaseInputTemplate = getTemplate<'BaseInputTemplate', T, S, F>('BaseInputTemplate', registry, options);
  const handleChange = useCallback((value: any) => onChange(value || undefined), [onChange]);

  return <BaseInputTemplate type='date' {...props} onChange={handleChange} />;
}
