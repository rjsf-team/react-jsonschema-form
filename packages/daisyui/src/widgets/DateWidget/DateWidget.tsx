import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { faCalendar } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { format, isValid, isToday, isSameDay } from 'date-fns';
import { DayPicker, ClassNames, ModifiersClassNames, UI } from 'react-day-picker';
import 'react-day-picker/dist/style.css';

import { FormContextType, RJSFSchema, StrictRJSFSchema, WidgetProps } from '@rjsf/utils';

//
// Types
//
interface DateTimePickerProps {
  selectedDate?: Date;
  month: Date;
  onMonthChange: (date: Date) => void;
  onSelect: (date: Date | undefined) => void;
}

//
// Hook to manage popup state and the displayed month
//
const useDatePickerState = (initialDate?: Date) => {
  const [isOpen, setIsOpen] = useState(false);
  const [month, setMonth] = useState<Date>(initialDate ?? new Date());
  return { isOpen, setIsOpen, month, setMonth };
};

//
// Hook for detecting clicks outside of a container
//
const useClickOutside = (ref: React.RefObject<HTMLDivElement>, callback: () => void) => {
  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (ref.current && !ref.current.contains(event.target as Node)) {
        callback();
      }
    };
    document.addEventListener('mousedown', handleClickOutside);
    return () => document.removeEventListener('mousedown', handleClickOutside);
  }, [ref, callback]);
};

//
// Predefined DayPicker styles using DaisyUI classes
//
const dayPickerStyles: { classNames: Partial<ClassNames>; modifiers: Partial<ModifiersClassNames> } = {
  classNames: {
    [UI.Root]: 'relative',
    [UI.Nav]: 'hidden',
    [UI.Chevron]: 'hidden',
    [UI.CaptionLabel]: 'hidden',
    [UI.Dropdowns]: 'flex justify-between gap-4 px-4 pb-4',
    [UI.Dropdown]: 'select select-bordered select-sm w-32',
    [UI.MonthsDropdown]: 'select select-bordered select-sm',
    [UI.YearsDropdown]: 'select select-bordered select-sm',
    [UI.Months]: 'flex justify-center',
    [UI.Month]: 'w-full',
    [UI.MonthCaption]: 'flex justify-center',
    [UI.MonthGrid]: 'w-full',
    [UI.Weekdays]: 'grid grid-cols-7 text-center border-b mb-2 pb-1 text-base-content/60 uppercase',
    [UI.Weekday]: 'p-1 font-medium text-base-content/60 text-sm',
    [UI.Week]: 'grid grid-cols-7',
    [UI.Day]: 'w-10 h-8 p-0 relative rounded-md',
    [UI.DayButton]:
      'btn btn-ghost absolute inset-0 flex items-center justify-center w-full h-full cursor-pointer rounded-md hover:btn-primary',
  },
  modifiers: {
    selected: 'btn btn-accent min-h-0 h-full',
    // today: 'btn btn-outline btn-info min-h-0 h-full',
    outside: 'text-base-content/30 hover:btn-ghost',
    disabled: 'opacity-50 cursor-not-allowed hover:btn-disabled',
  },
};

//
// Popup component for the calendar.
// Wrapped with React.memo to avoid unnecessary reâ€‘renders.
//
const DateTimePickerPopup: React.FC<DateTimePickerProps> = React.memo(
  ({ selectedDate, month, onMonthChange, onSelect }) => {
    const customDayModifiers = {
      selected: selectedDate, // DayPicker uses this for tracking selection.
      'custom-today': (date: Date) => isToday(date) && !(selectedDate && isSameDay(date, selectedDate)),
    };

    const customModifiersClassNames: ModifiersClassNames = {
      selected: dayPickerStyles.modifiers.selected as string,
      'custom-today': 'btn btn-outline btn-info min-h-0 h-full',
    };

    return (
      <div>
        <DayPicker
          captionLayout='dropdown'
          classNames={dayPickerStyles.classNames}
          fromYear={1800}
          toYear={2025}
          mode='single'
          modifiers={customDayModifiers}
          modifiersClassNames={customModifiersClassNames}
          month={month}
          onMonthChange={onMonthChange}
          onSelect={onSelect}
          selected={selectedDate}
          showOutsideDays
        />
      </div>
    );
  }
);

//
// Main widget component
//
const DateTimeWidget = <T = any, S extends StrictRJSFSchema = RJSFSchema, F extends FormContextType = any>(
  props: WidgetProps<T, S, F>
) => {
  const { id, value, onChange, schema } = props;
  // Initialize the local date from the parent's value.
  const initialDate = useMemo(() => (value ? new Date(value) : undefined), [value]);
  const [localDate, setLocalDate] = useState<Date | undefined>(initialDate);

  // When the parent's value changes externally, update local state.
  useEffect(() => {
    setLocalDate(initialDate);
  }, [initialDate]);

  const { isOpen, setIsOpen, month, setMonth } = useDatePickerState(initialDate);
  const containerRef = useRef<HTMLDivElement>(null);

  // Close the popup when clicking outside and commit changes.
  useClickOutside(containerRef, () => {
    setIsOpen(false);
    onChange(localDate ? localDate.toISOString() : '');
  });

  // When the local date changes, update the displayed month.
  useEffect(() => {
    if (localDate) {
      setMonth(localDate);
    }
  }, [localDate, setMonth]);

  // Update the month when the user navigates the calendar.
  const handleMonthChange = useCallback((date: Date) => setMonth(date), [setMonth]);

  // Update local state on day selection (but do not commit immediately).
  const handleSelect = useCallback((date: Date | undefined) => {
    if (date) {
      // Remove any time component by setting hours, minutes, seconds, and milliseconds to zero.
      date.setHours(0, 0, 0, 0);
      setLocalDate(date);
    }
  }, []);

  // Toggle popup visibility.
  const togglePicker = useCallback(
    (e: React.MouseEvent) => {
      e.stopPropagation();
      setIsOpen((prev) => !prev);
    },
    [setIsOpen]
  );

  return (
    <div ref={containerRef} className='form-control my-4 w-full relative'>
      <div
        className='w-full'
        tabIndex={0}
        onKeyDown={(e) => {
          if (e.key === 'Enter' || e.key === ' ') {
            togglePicker(e as unknown as React.MouseEvent);
          }
        }}
      >
        <div
          id={id}
          className={`input input-bordered w-full flex items-center justify-between cursor-pointer ${
            isOpen ? 'ring-2 ring-primary/50' : ''
          }`}
          onClick={togglePicker}
          role='button'
          aria-haspopup='true'
          aria-expanded={isOpen}
          tabIndex={-1}
        >
          <span className={localDate && isValid(localDate) ? '' : 'text-base-content/50'}>
            {localDate && isValid(localDate) ? format(localDate, 'PP') : schema.title}
          </span>
          <FontAwesomeIcon icon={faCalendar} className='ml-2 h-4 w-4 text-primary-content' />
        </div>
        {isOpen && (
          <div className='absolute z-50 mt-2 w-full max-w-xs bg-base-100 border border-base-300 shadow-lg rounded-box p-4'>
            <DateTimePickerPopup
              selectedDate={localDate}
              month={month}
              onMonthChange={handleMonthChange}
              onSelect={handleSelect}
            />
            <div className='mt-2 flex justify-end'>
              <button
                type='button'
                className='btn btn-sm btn-primary'
                onClick={() => {
                  setIsOpen(false);
                  onChange(localDate ? localDate.toISOString() : '');
                }}
              >
                Done
              </button>
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default DateTimeWidget;
