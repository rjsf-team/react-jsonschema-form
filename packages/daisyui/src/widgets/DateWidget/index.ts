export { default } from './DateWidget';
export * from './DateWidget';
