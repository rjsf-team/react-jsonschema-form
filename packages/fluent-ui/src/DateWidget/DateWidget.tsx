import React from "react";
import { WidgetProps } from "@rjsf/core";
import {
  DatePicker,
  DayOfWeek,
  IDatePickerStrings,
  mergeStyleSets,
} from "@fluentui/react";
import _pick from "lodash/pick";
import { utils } from "@rjsf/core";

const { pad } = utils;

// Keys of IDropdownProps from @fluentui/react
const allowedProps = [
  "componentRef",
  "styles",
  "theme",
  "calloutProps",
  "calendarProps",
  "textField",
  "calendarAs",
  "onSelectDate",
  "label",
  "isRequired",
  "disabled",
  "ariaLabel",
  "underlined",
  "pickerAriaLabel",
  "isMonthPickerVisible",
  "showMonthPickerAsOverlay",
  "allowTextInput",
  "disableAutoFocus",
  "placeholder",
  "today",
  "value",
  "formatDate",
  "parseDateFromString",
  "firstDayOfWeek",
  "strings",
  "highlightCurrentMonth",
  "highlightSelectedMonth",
  "showWeekNumbers",
  "firstWeekOfYear",
  "showGoToToday",
  "borderless",
  "className",
  "dateTimeFormatter",
  "minDate",
  "maxDate",
  "initialPickerDate",
  "allFocusable",
  "onAfterMenuDismiss",
  "showCloseButton",
  "tabIndex"
];

const controlClass = mergeStyleSets({
  control: {
    margin: "0 0 15px 0",
  },
});

// TODO: move to utils.
// TODO: figure out a standard format for this, as well as
// how we can get this to work with locales.
const formatDate = (date?: Date) => {
  if (!date) {
    return "";
  }
  const yyyy = pad(date.getFullYear(), 4);	
  const MM = pad(date.getMonth() + 1, 2);	
  const dd = pad(date.getDate(), 2);
  return `${yyyy}-${MM}-${dd}`
}

const parseDate = (dateStr?: string) => {
  if (!dateStr) {
    return undefined;
  }
  const [year, month, day] = dateStr.split("-").map(e => parseInt(e));
  var dt = new Date(year, month - 1, day);
  return dt;
}

const DateWidget = ({
  id,
  required,
  readonly,
  disabled,
  label,
  value,
  onChange,
  onBlur,
  onFocus,
  autofocus,
  options,
  placeholder,
  schema,
  rawErrors,
}: WidgetProps) => {
  const _onSelectDate = (date: Date | null | undefined) => {
    if (date) {
      const formatted = formatDate(date);
      formatted && onChange(formatted);
    }
  }
  const _onBlur = ({ target: { value } }: React.FocusEvent<HTMLInputElement>) =>
    onBlur(id, value);
  const _onFocus = ({
    target: { value },
  }: React.FocusEvent<HTMLInputElement>) => onFocus(id, value);

  const uiProps = _pick(options.props || {}, allowedProps);
  return (
    <DatePicker
      className={controlClass.control}
      firstDayOfWeek={DayOfWeek.Sunday}
      placeholder={placeholder}
      ariaLabel="Select a date"
      isRequired={required}
      label={label}
      onSelectDate={_onSelectDate}
      onBlur={_onBlur}
      onFocus={_onFocus}
      value={parseDate(value)}
      {...uiProps}
    />
  );
};

export default DateWidget;
