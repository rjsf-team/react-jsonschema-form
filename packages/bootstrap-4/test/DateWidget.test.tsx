import React from "react";
import DateWidget from "../src/DateWidget";
import renderer from "react-test-renderer";
import { makeWidgetMockProps } from "./helpers/createMocks";

describe("DateWidget", () => {
  test("simple", () => {
    const tree = renderer
      .create(<DateWidget {...makeWidgetMockProps({})} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
