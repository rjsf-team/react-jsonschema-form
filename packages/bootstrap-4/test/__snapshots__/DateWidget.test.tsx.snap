// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateWidget simple 1`] = `
<div
  className="mb-0 form-group"
>
  <label
    className="text-danger form-label"
  >
    Some simple label
    *
  </label>
  <input
    autoFocus={true}
    className="is-invalid form-control"
    disabled={false}
    id="_id"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    readOnly={true}
    required={true}
    type="date"
    value="value"
  />
</div>
`;
